File: python/mirascope/ops/_internal/instrumentation/providers/openai.py
"""OpenTelemetry instrumentation for OpenAI SDK."""

from __future__ import annotations

from typing import TYPE_CHECKING

from opentelemetry.instrumentation.openai_v2 import OpenAIInstrumentor

from .base import BaseInstrumentation, ContentCaptureMode

if TYPE_CHECKING:
    from opentelemetry.trace import TracerProvider


class OpenAIInstrumentation(BaseInstrumentation[OpenAIInstrumentor]):
    """Manages OpenTelemetry instrumentation lifecycle for the OpenAI SDK."""

    def _create_instrumentor(self) -> OpenAIInstrumentor:
        """Create a new OpenAI instrumentor instance."""
        return OpenAIInstrumentor()

    def _configure_capture_content(self, capture_content: ContentCaptureMode) -> None:
        """Configure environment variables for OpenAI content capture."""
        if capture_content == "enabled":
            self._set_env_var(
                "OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT", "true"
            )
        elif capture_content == "disabled":
            self._set_env_var(
                "OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT", "false"
            )


def instrument_openai(
    *,
    tracer_provider: TracerProvider | None = None,
    capture_content: ContentCaptureMode = "default",
) -> None:
    """Enable OpenTelemetry instrumentation for the OpenAI Python SDK.

    Uses the provided tracer_provider or the global OpenTelemetry tracer provider.

    Args:
        tracer_provider: Optional tracer provider to use. If not provided,
            uses the global OpenTelemetry tracer provider.
        capture_content: Controls whether to capture message content in spans.
            - "enabled": Capture message content
            - "disabled": Do not capture message content
            - "default": Use the library's default behavior

    Example:

        Enable instrumentation with Mirascope Cloud:
        ```python
        from mirascope import ops

        ops.configure()
        ops.instrument_openai()
        ```

        Enable instrumentation with content capture:
        ```python
        from mirascope import ops

        ops.configure()
        ops.instrument_openai(capture_content="enabled")
        ```
    """
    OpenAIInstrumentation().instrument(
        tracer_provider=tracer_provider,
        capture_content=capture_content,
    )


def uninstrument_openai() -> None:
    """Disable previously configured OpenAI instrumentation."""
    OpenAIInstrumentation().uninstrument()


def is_openai_instrumented() -> bool:
    """Return whether OpenAI instrumentation is currently active."""
    return OpenAIInstrumentation().is_instrumented

File: typescript/src/ops/_internal/instrumentation/providers/openai.ts
/**
 * OpenTelemetry instrumentation for OpenAI SDK.
 *
 * Uses the official @opentelemetry/instrumentation-openai package.
 */

import type { TracerProvider } from "@opentelemetry/api";

import {
  BaseInstrumentation,
  type ContentCaptureMode,
  type Instrumentor,
} from "@/ops/_internal/instrumentation/providers/base";

// Lazy import to avoid requiring the package if not used
let OpenAIInstrumentationClass: (new () => Instrumentor) | null = null;

/* v8 ignore start - requires optional dependency */
async function getOpenAIInstrumentationClass(): Promise<
  new () => Instrumentor
> {
  if (OpenAIInstrumentationClass) {
    return OpenAIInstrumentationClass;
  }

  try {
    // Dynamic import of optional dependency - uses type declaration from types.d.ts
    const { OpenAIInstrumentation } =
      await import("@opentelemetry/instrumentation-openai");
    OpenAIInstrumentationClass =
      OpenAIInstrumentation as new () => Instrumentor;
    return OpenAIInstrumentationClass;
  } catch {
    throw new Error(
      "Failed to import @opentelemetry/instrumentation-openai. " +
        "Please install it: npm install @opentelemetry/instrumentation-openai",
    );
  }
}
/* v8 ignore end */

// Singleton state
let openaiInstance: OpenAIInstrumentationManager | null = null;
let openaiInstrumentorClass: (new () => Instrumentor) | null = null;

/**
 * Manages OpenTelemetry instrumentation lifecycle for the OpenAI SDK.
 */
class OpenAIInstrumentationManager extends BaseInstrumentation<Instrumentor> {
  static instance(): OpenAIInstrumentationManager {
    if (!openaiInstance) {
      openaiInstance = new OpenAIInstrumentationManager();
    }
    return openaiInstance;
  }

  protected createInstrumentor(): Instrumentor {
    if (!openaiInstrumentorClass) {
      throw new Error(
        "Instrumentor class not loaded. Call instrumentOpenai() instead of using the class directly.",
      );
    }
    return new openaiInstrumentorClass();
  }

  protected configureContentCapture(captureContent: ContentCaptureMode): void {
    if (captureContent === "enabled") {
      this.setEnvVar(
        "OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT",
        "true",
      );
    } else if (captureContent === "disabled") {
      this.setEnvVar(
        "OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT",
        "false",
      );
    }
  }

  static resetForTesting(): void {
    if (openaiInstance) {
      openaiInstance.uninstrument();
    }
    openaiInstance = null;
    openaiInstrumentorClass = null;
  }
}

/**
 * Enable OpenTelemetry instrumentation for the OpenAI SDK.
 *
 * Uses the provided tracerProvider or the global OpenTelemetry tracer provider.
 *
 * @param options - Configuration options
 * @param options.tracerProvider - Optional tracer provider to use
 * @param options.captureContent - Controls whether to capture message content in spans
 *
 * @example
 * ```typescript
 * import { ops } from '@mirascope/core';
 *
 * // Configure first
 * ops.configure();
 *
 * // Enable OpenAI instrumentation
 * await ops.instrumentOpenai();
 *
 * // With content capture enabled
 * await ops.instrumentOpenai({ captureContent: "enabled" });
 * ```
 */
/* v8 ignore start - requires optional dependency */
export async function instrumentOpenai(
  options: {
    tracerProvider?: TracerProvider;
    captureContent?: ContentCaptureMode;
  } = {},
): Promise<void> {
  openaiInstrumentorClass = await getOpenAIInstrumentationClass();
  const manager = OpenAIInstrumentationManager.instance();
  manager.instrument(options);
}
/* v8 ignore end */

/**
 * Disable previously configured OpenAI instrumentation.
 */
export function uninstrumentOpenai(): void {
  OpenAIInstrumentationManager.instance().uninstrument();
}

/**
 * Return whether OpenAI instrumentation is currently active.
 */
export function isOpenaiInstrumented(): boolean {
  return OpenAIInstrumentationManager.instance().isInstrumented;
}

/**
 * Reset OpenAI instrumentation for testing.
 */
export function resetOpenaiInstrumentationForTesting(): void {
  OpenAIInstrumentationManager.resetForTesting();
}
